ipie/addons/adafqmc/qmc/adafqmc.py

```diff
@@ -0,0 +1,493 @@
+import numpy as np
+import torch
+from torch.utils.checkpoint import checkpoint
+from ipie.addons.adafqmc.hamiltonians.hamiltonian import ham_with_obs, rot_ham_with_orbs
+from ipie.addons.adafqmc.trial_wavefunction.sdtrial import SDTrial
+from ipie.addons.adafqmc.walkers.rhf_walkers import (
+    Walkers,
+    initialize_walkers,
+    reorthogonalize,
+    stochastic_reconfiguration,
+)
+from ipie.addons.adafqmc.propagation.propagator import Propagator
+import time
+
+
+class QMCParams:
+    def __init__(
+        self,
+        num_walkers_per_process,
+        num_steps_per_block,
+        ad_block_size,
+        num_ad_blocks,
+        timestep,
+        stabilize_freq,
+        pop_control_freq,
+        pop_control_freq_eq,
+        seed,
+        grad_checkpointing=False,
+        chkpt_size=50,
+    ):
+        """Initialize
+
+        Parameters
+        -------
+        num_walkers_per_process : int
+            Number of walkers per process
+        num_steps_per_block : int
+            Number of steps per AD block
+        num_ad_blocks : int
+            Number of AD blocks
+        """
+        self.num_walkers_per_process = num_walkers_per_process
+        self.num_steps_per_block = num_steps_per_block
+        self.ad_block_size = ad_block_size
+        self.num_ad_blocks = num_ad_blocks
+        self.timestep = timestep
+        self.stabilize_freq = stabilize_freq
+        self.pop_control_freq = pop_control_freq
+        self.pop_control_freq_eq = pop_control_freq_eq
+        self.seed = seed
+        self.grad_checkpointing = grad_checkpointing
+        self.chkpt_size = chkpt_size
+        if self.grad_checkpointing and num_steps_per_block % chkpt_size != 0:
+            raise ValueError(
+                "The number of steps per block should be divisible by the checkpoint size"
+            )
+
+
+class ADAFQMC:
+    def __init__(self, comm, trial_tangent: callable, params: QMCParams):
+        """Initialize
+
+        Parameters
+        -------
+        comm : MPI communicator
+        trial_tangent : function
+            function that stores the trial and the tangent of the trial wavefunction
+        params : Dist_Params
+            parameters for the distributed AFQMC
+        """
+        # trial and walkers will be generated afterwards
+        # specify the set_trial function with given tangent
+        self.comm = comm
+        self.rank = comm.Get_rank()
+        self.size = comm.Get_size()
+        self.set_trial = trial_tangent
+        self.params = params
+
+    @staticmethod
+    def build(
+        comm,
+        trial_tangent: callable,  # specify the function that obtains the trial wavefunction with precomputed trial and tangent
```

Comment:
Sorry, it should be like: https://docs.python.org/3/library/typing.html#annotating-callable-objects

Reply:
Oh my fault. I'll change